File: jas_ocaml/test/algorithms/planar_test.ml
(** Planar graph extraction tests. Mirrors the Rust suite at
    jas_dioxus/src/algorithms/planar.rs. *)

open Jas.Planar

let area_eps = 1e-6

let approx_eq a b = abs_float (a -. b) < area_eps

let closed_square x y side : polyline =
  [| (x, y);
     (x +. side, y);
     (x +. side, y +. side);
     (x, y +. side);
     (x, y) |]

let segment a b : polyline = [| a; b |]

let total_top_level_area g =
  List.fold_left
    (fun acc f -> acc +. abs_float (face_net_area g f))
    0.0 (top_level_faces g)

let check name cond =
  if not cond then failwith ("FAIL: " ^ name)

(* ----- 1. Two crossing segments ----- *)
let test_two_crossing_segments_have_no_bounded_faces () =
  let g = build [
    segment (-1.0, 0.0) (1.0, 0.0);
    segment (0.0, -1.0) (0.0, 1.0);
  ] in
  check "two_crossing.face_count" (face_count g = 0)

(* ----- 2. Closed square ----- *)
let test_closed_square_is_one_face () =
  let g = build [closed_square 0.0 0.0 10.0] in
  check "closed_square.face_count" (face_count g = 1);
  check "closed_square.area" (approx_eq (total_top_level_area g) 100.0)

(* ----- 3. Square with one diagonal ----- *)
let test_square_with_one_diagonal () =
  let g = build [
    closed_square 0.0 0.0 10.0;
    segment (0.0, 0.0) (10.0, 10.0);
  ] in
  check "one_diag.face_count" (face_count g = 2);
  check "one_diag.area" (approx_eq (total_top_level_area g) 100.0);
  List.iter (fun f ->
    check "one_diag.tri_area" (approx_eq (abs_float (face_net_area g f)) 50.0)
  ) (top_level_faces g)

(* ----- 4. Square with both diagonals ----- *)
let test_square_with_both_diagonals () =
  let g = build [
    closed_square 0.0 0.0 10.0;
    segment (0.0, 0.0) (10.0, 10.0);
    segment (10.0, 0.0) (0.0, 10.0);
  ] in
  check "both_diag.face_count" (face_count g = 4);
  check "both_diag.area" (approx_eq (total_top_level_area g) 100.0);
  List.iter (fun f ->
    check "both_diag.tri_area" (approx_eq (abs_float (face_net_area g f)) 25.0)
  ) (top_level_faces g)

(* ----- 5. Two disjoint squares ----- *)
let test_two_disjoint_squares () =
  let g = build [
    closed_square 0.0 0.0 10.0;
    closed_square 20.0 0.0 10.0;
  ] in
  check "disjoint.face_count" (face_count g = 2);
  check "disjoint.area" (approx_eq (total_top_level_area g) 200.0)

(* ----- 6. Two squares sharing an edge ----- *)
let test_two_squares_sharing_an_edge () =
  let g = build [
    closed_square 0.0 0.0 10.0;
    closed_square 10.0 0.0 10.0;
  ] in
  check "shared.face_count" (face_count g = 2);
  check "shared.area" (approx_eq (total_top_level_area g) 200.0)

(* ----- 8. Concentric squares ----- *)
let test_square_with_inner_square () =
  let g = build [
    closed_square 0.0 0.0 20.0;
    closed_square 5.0 5.0 10.0;
  ] in
  check "concentric.face_count" (face_count g = 2);
  let top = top_level_faces g in
  check "concentric.top_count" (List.length top = 1);
  let outer = List.hd top in
  check "concentric.holes_count"
    (List.length g.faces.(outer).holes = 1);
  check "concentric.outer_area"
    (approx_eq (abs_float (face_outer_area g outer)) 400.0);
  check "concentric.net_area"
    (approx_eq (abs_float (face_net_area g outer)) 300.0);
  let inner = ref (-1) in
  for i = 0 to Array.length g.faces - 1 do
    if g.faces.(i).depth = 2 then inner := i
  done;
  check "concentric.inner_found" (!inner >= 0);
  check "concentric.inner_parent" (g.faces.(!inner).parent = Some outer);
  check "concentric.inner_area"
    (approx_eq (abs_float (face_net_area g !inner)) 100.0)

(* ----- 9. Hit test on the diagonal-cross square ----- *)
let test_hit_test_diagonal_quadrants () =
  let g = build [
    closed_square 0.0 0.0 10.0;
    segment (0.0, 0.0) (10.0, 10.0);
    segment (10.0, 0.0) (0.0, 10.0);
  ] in
  let samples = [
    (5.0, 1.0); (9.0, 5.0); (5.0, 9.0); (1.0, 5.0);
  ] in
  let hits = List.map (fun s ->
    match hit_test g s with
    | Some f -> f
    | None -> failwith "expected hit"
  ) samples in
  let unique = List.sort_uniq compare hits in
  check "diag_hit.distinct" (List.length unique = 4)

(* ----- 10. Degenerate inputs ----- *)
let test_empty_input () =
  let g = build [] in
  check "empty.face_count" (face_count g = 0)

let test_zero_length_segment () =
  let g = build [segment (1.0, 1.0) (1.0, 1.0)] in
  check "zero.face_count" (face_count g = 0)

let test_single_point_polyline () =
  let g = build [[| (3.0, 3.0) |]] in
  check "single.face_count" (face_count g = 0)

(* ----- 11. Square with external tail ----- *)
let test_square_with_external_tail () =
  let g = build [
    closed_square 0.0 0.0 10.0;
    segment (10.0, 10.0) (15.0, 15.0);
  ] in
  check "ext_tail.face_count" (face_count g = 1);
  check "ext_tail.area" (approx_eq (total_top_level_area g) 100.0)

(* ----- 12. Square with internal tail ----- *)
let test_square_with_internal_tail () =
  let g = build [
    closed_square 0.0 0.0 10.0;
    segment (0.0, 0.0) (5.0, 5.0);
  ] in
  check "int_tail.face_count" (face_count g = 1);
  check "int_tail.area" (approx_eq (total_top_level_area g) 100.0)

(* ----- 13. Square with branching tree ----- *)
let test_square_with_internal_tree () =
  let g = build [
    closed_square 0.0 0.0 10.0;
    [| (0.0, 0.0); (3.0, 3.0) |];
    [| (3.0, 3.0); (5.0, 3.0) |];
    [| (3.0, 3.0); (3.0, 5.0) |];
    [| (5.0, 3.0); (6.0, 4.0) |];
  ] in
  check "tree.face_count" (face_count g = 1);
  check "tree.area" (approx_eq (total_top_level_area g) 100.0)

(* ----- 14. Isolated open stroke ----- *)
let test_isolated_open_stroke () =
  let g = build [segment (0.0, 0.0) (5.0, 5.0)] in
  check "isolated.face_count" (face_count g = 0)

(* ----- 15. Square with two disjoint holes ----- *)
let test_square_with_two_disjoint_holes () =
  let g = build [
    closed_square 0.0 0.0 30.0;
    closed_square 5.0 5.0 5.0;
    closed_square 20.0 20.0 5.0;
  ] in
  check "two_holes.face_count" (face_count g = 3);
  let top = top_level_faces g in
  check "two_holes.top_count" (List.length top = 1);
  let outer = List.hd top in
  check "two_holes.holes_count"
    (List.length g.faces.(outer).holes = 2);
  check "two_holes.net_area"
    (approx_eq (abs_float (face_net_area g outer)) 850.0)

(* ----- 16. Three-deep nested squares ----- *)
let test_three_deep_nested () =
  let g = build [
    closed_square 0.0 0.0 30.0;
    closed_square 5.0 5.0 20.0;
    closed_square 10.0 10.0 10.0;
  ] in
  check "nested.face_count" (face_count g = 3);
  let by_depth = Array.make 4 [] in
  for i = 0 to Array.length g.faces - 1 do
    let d = g.faces.(i).depth in
    if d >= 1 && d <= 3 then by_depth.(d) <- i :: by_depth.(d)
  done;
  check "nested.depth1" (List.length by_depth.(1) = 1);
  check "nested.depth2" (List.length by_depth.(2) = 1);
  check "nested.depth3" (List.length by_depth.(3) = 1);
  let a = List.hd by_depth.(1) in
  let b = List.hd by_depth.(2) in
  let c = List.hd by_depth.(3) in
  check "nested.b_parent" (g.faces.(b).parent = Some a);
  check "nested.c_parent" (g.faces.(c).parent = Some b);
  check "nested.a_area" (approx_eq (abs_float (face_net_area g a)) 500.0);
  check "nested.b_area" (approx_eq (abs_float (face_net_area g b)) 300.0);
  check "nested.c_area" (approx_eq (abs_float (face_net_area g c)) 100.0)

(* ----- 17. Hit test inside a hole ----- *)
let test_hit_test_in_hole () =
  let g = build [
    closed_square 0.0 0.0 20.0;
    closed_square 5.0 5.0 10.0;
  ] in
  let outer_hit =
    match hit_test g (1.0, 1.0) with
    | Some f -> f
    | None -> failwith "outer expected"
  in
  check "hole_hit.outer_depth" (g.faces.(outer_hit).depth = 1);
  let hole_hit =
    match hit_test g (10.0, 10.0) with
    | Some f -> f
    | None -> failwith "hole expected"
  in
  check "hole_hit.hole_depth" (g.faces.(hole_hit).depth = 2);
  check "hole_hit.hole_parent" (g.faces.(hole_hit).parent = Some outer_hit)

let () =
  Alcotest.run "Planar" [
    "planar", [
      Alcotest.test_case "two_crossing_segments_have_no_bounded_faces" `Quick test_two_crossing_segments_have_no_bounded_faces;
      Alcotest.test_case "closed_square_is_one_face" `Quick test_closed_square_is_one_face;
      Alcotest.test_case "square_with_one_diagonal" `Quick test_square_with_one_diagonal;
      Alcotest.test_case "square_with_both_diagonals" `Quick test_square_with_both_diagonals;
      Alcotest.test_case "two_disjoint_squares" `Quick test_two_disjoint_squares;
      Alcotest.test_case "two_squares_sharing_an_edge" `Quick test_two_squares_sharing_an_edge;
      Alcotest.test_case "square_with_inner_square" `Quick test_square_with_inner_square;
      Alcotest.test_case "hit_test_diagonal_quadrants" `Quick test_hit_test_diagonal_quadrants;
      Alcotest.test_case "empty_input" `Quick test_empty_input;
      Alcotest.test_case "zero_length_segment" `Quick test_zero_length_segment;
      Alcotest.test_case "single_point_polyline" `Quick test_single_point_polyline;
      Alcotest.test_case "square_with_external_tail" `Quick test_square_with_external_tail;
      Alcotest.test_case "square_with_internal_tail" `Quick test_square_with_internal_tail;
      Alcotest.test_case "square_with_internal_tree" `Quick test_square_with_internal_tree;
      Alcotest.test_case "isolated_open_stroke" `Quick test_isolated_open_stroke;
      Alcotest.test_case "square_with_two_disjoint_holes" `Quick test_square_with_two_disjoint_holes;
      Alcotest.test_case "three_deep_nested" `Quick test_three_deep_nested;
      Alcotest.test_case "hit_test_in_hole" `Quick test_hit_test_in_hole;
    ];
  ]

File: jas/algorithms/planar_test.py
"""Planar graph extraction tests. Mirrors the Rust suite at
jas_dioxus/src/algorithms/planar.rs."""

from __future__ import annotations

import pytest

from algorithms.planar import build


AREA_EPS = 1e-6


def closed_square(x, y, side):
    return [
        (x, y),
        (x + side, y),
        (x + side, y + side),
        (x, y + side),
        (x, y),
    ]


def segment(a, b):
    return [a, b]


def total_top_level_area(g):
    return sum(abs(g.face_net_area(f)) for f in g.top_level_faces())


# ----- 1. Two crossing segments -----

def test_two_crossing_segments_have_no_bounded_faces():
    g = build([
        segment((-1.0, 0.0), (1.0, 0.0)),
        segment((0.0, -1.0), (0.0, 1.0)),
    ])
    assert g.face_count() == 0


# ----- 2. Closed square -----

def test_closed_square_is_one_face():
    g = build([closed_square(0.0, 0.0, 10.0)])
    assert g.face_count() == 1
    assert abs(total_top_level_area(g) - 100.0) < AREA_EPS


# ----- 3. Square with one diagonal -----

def test_square_with_one_diagonal():
    g = build([
        closed_square(0.0, 0.0, 10.0),
        segment((0.0, 0.0), (10.0, 10.0)),
    ])
    assert g.face_count() == 2
    assert abs(total_top_level_area(g) - 100.0) < AREA_EPS
    for f in g.top_level_faces():
        assert abs(abs(g.face_net_area(f)) - 50.0) < AREA_EPS


# ----- 4. Square with both diagonals -----

def test_square_with_both_diagonals():
    g = build([
        closed_square(0.0, 0.0, 10.0),
        segment((0.0, 0.0), (10.0, 10.0)),
        segment((10.0, 0.0), (0.0, 10.0)),
    ])
    assert g.face_count() == 4
    assert abs(total_top_level_area(g) - 100.0) < AREA_EPS
    for f in g.top_level_faces():
        assert abs(abs(g.face_net_area(f)) - 25.0) < AREA_EPS


# ----- 5. Two disjoint squares -----

def test_two_disjoint_squares():
    g = build([
        closed_square(0.0, 0.0, 10.0),
        closed_square(20.0, 0.0, 10.0),
    ])
    assert g.face_count() == 2
    assert abs(total_top_level_area(g) - 200.0) < AREA_EPS


# ----- 6. Two squares sharing an edge -----

def test_two_squares_sharing_an_edge():
    g = build([
        closed_square(0.0, 0.0, 10.0),
        closed_square(10.0, 0.0, 10.0),
    ])
    assert g.face_count() == 2
    assert abs(total_top_level_area(g) - 200.0) < AREA_EPS


# ----- 7. T-junction (deferred) -----

@pytest.mark.skip(reason="T-junctions where one polyline's vertex lands on another's interior not yet supported")
def test_t_junction():
    g = build([
        segment((0.0, 0.0), (10.0, 0.0)),
        segment((5.0, 0.0), (5.0, 5.0)),
    ])
    assert g.face_count() == 0


# ----- 8. Concentric squares -----

def test_square_with_inner_square():
    g = build([
        closed_square(0.0, 0.0, 20.0),
        closed_square(5.0, 5.0, 10.0),
    ])
    assert g.face_count() == 2
    top = g.top_level_faces()
    assert len(top) == 1
    outer = top[0]
    assert len(g.faces[outer].holes) == 1
    assert abs(abs(g.face_outer_area(outer)) - 400.0) < AREA_EPS
    assert abs(abs(g.face_net_area(outer)) - 300.0) < AREA_EPS
    inner = next(
        i for i, f in enumerate(g.faces) if f.depth == 2
    )
    assert g.faces[inner].parent == outer
    assert abs(abs(g.face_net_area(inner)) - 100.0) < AREA_EPS


# ----- 9. Hit test on the diagonal-cross square -----

def test_hit_test_diagonal_quadrants():
    g = build([
        closed_square(0.0, 0.0, 10.0),
        segment((0.0, 0.0), (10.0, 10.0)),
        segment((10.0, 0.0), (0.0, 10.0)),
    ])
    samples = [(5.0, 1.0), (9.0, 5.0), (5.0, 9.0), (1.0, 5.0)]
    hits = []
    for s in samples:
        f = g.hit_test(s)
        assert f is not None
        hits.append(f)
    assert len(set(hits)) == 4


# ----- 10. Degenerate inputs -----

def test_empty_input():
    g = build([])
    assert g.face_count() == 0


def test_zero_length_segment():
    g = build([segment((1.0, 1.0), (1.0, 1.0))])
    assert g.face_count() == 0


def test_single_point_polyline():
    g = build([[(3.0, 3.0)]])
    assert g.face_count() == 0


# ----- 11. Square with external tail -----

def test_square_with_external_tail():
    g = build([
        closed_square(0.0, 0.0, 10.0),
        segment((10.0, 10.0), (15.0, 15.0)),
    ])
    assert g.face_count() == 1
    assert abs(total_top_level_area(g) - 100.0) < AREA_EPS


# ----- 12. Square with internal tail -----

def test_square_with_internal_tail():
    g = build([
        closed_square(0.0, 0.0, 10.0),
        segment((0.0, 0.0), (5.0, 5.0)),
    ])
    assert g.face_count() == 1
    assert abs(total_top_level_area(g) - 100.0) < AREA_EPS


# ----- 13. Square with branching tree -----

def test_square_with_internal_tree():
    g = build([
        closed_square(0.0, 0.0, 10.0),
        [(0.0, 0.0), (3.0, 3.0)],
        [(3.0, 3.0), (5.0, 3.0)],
        [(3.0, 3.0), (3.0, 5.0)],
        [(5.0, 3.0), (6.0, 4.0)],
    ])
    assert g.face_count() == 1
    assert abs(total_top_level_area(g) - 100.0) < AREA_EPS


# ----- 14. Isolated open stroke -----

def test_isolated_open_stroke():
    g = build([segment((0.0, 0.0), (5.0, 5.0))])
    assert g.face_count() == 0


# ----- 15. Square with two disjoint holes -----

def test_square_with_two_disjoint_holes():
    g = build([
        closed_square(0.0, 0.0, 30.0),
        closed_square(5.0, 5.0, 5.0),
        closed_square(20.0, 20.0, 5.0),
    ])
    assert g.face_count() == 3
    top = g.top_level_faces()
    assert len(top) == 1
    outer = top[0]
    assert len(g.faces[outer].holes) == 2
    assert abs(abs(g.face_net_area(outer)) - 850.0) < AREA_EPS


# ----- 16. Three-deep nested squares -----

def test_three_deep_nested():
    g = build([
        closed_square(0.0, 0.0, 30.0),
        closed_square(5.0, 5.0, 20.0),
        closed_square(10.0, 10.0, 10.0),
    ])
    assert g.face_count() == 3
    by_depth: dict[int, list[int]] = {}
    for i, f in enumerate(g.faces):
        by_depth.setdefault(f.depth, []).append(i)
    assert len(by_depth.get(1, [])) == 1
    assert len(by_depth.get(2, [])) == 1
    assert len(by_depth.get(3, [])) == 1
    a = by_depth[1][0]
    b = by_depth[2][0]
    c = by_depth[3][0]
    assert g.faces[b].parent == a
    assert g.faces[c].parent == b
    assert abs(abs(g.face_net_area(a)) - 500.0) < AREA_EPS
    assert abs(abs(g.face_net_area(b)) - 300.0) < AREA_EPS
    assert abs(abs(g.face_net_area(c)) - 100.0) < AREA_EPS


# ----- 17. Hit test inside a hole -----

def test_hit_test_in_hole():
    g = build([
        closed_square(0.0, 0.0, 20.0),
        closed_square(5.0, 5.0, 10.0),
    ])
    outer_hit = g.hit_test((1.0, 1.0))
    assert outer_hit is not None
    assert g.faces[outer_hit].depth == 1
    hole_hit = g.hit_test((10.0, 10.0))
    assert hole_hit is not None
    assert g.faces[hole_hit].depth == 2
    assert g.faces[hole_hit].parent == outer_hit


# ----- Deferred / known limitations -----

@pytest.mark.skip(reason="collinear self-overlap not yet supported (mirrors boolean_normalize)")
def test_collinear_overlap():
    pass


@pytest.mark.skip(reason="incremental rebuild not yet supported")
def test_incremental_add_stroke():
    pass
